### PR TITLE
Addition of autoencoder class. 

### PR DESCRIPTION
**Pull request by Ljackson-su**
Capability of working with video and images, Encode and decode reproduces the same video. Not yet tested in video diffusion pipeline

### CODE REVIEW

File: mindglass_code/autoencoder.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    # inputs: (b, dh, dw)
+    # outputs: (b, 4, dh*8, dw*8)
+    def decode_img_latents(self, latents):
+        latents = 1 / 0.18215 * latents
+
+        with torch.no_grad():
+            imgs = self.model.decode(latents)
+
+        # Unsure of resizing but matches that in the encoding
+        imgs = (imgs[0] / 2 + 0.5).clamp(0, 1)
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:58]: Are you sure you need the 0-index here? 
imgs is of size [b, c, h, w] and imgs[0] would essentially take the first batch which has shape [c, h, w].
If you want to apply the de-normalization here, I believe it should be:
`imgs = (imgs / 2 + 0.5).clamp(0, 1)`

File: mindglass_code/autoencoder.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        imgs = imgs.detach().cpu().permute(0, 2, 3, 1).numpy()
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:59]: This does not resize, it just permutes the order, from [b, c, h, w] to [b, h, w, c].

File: mindglass_code/autoencoder.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    # Define function that encodes a video
+    # Encodes videos to the latent space.
+    # Can pass in videos of any size - all dims recuded by a factor of 8
+    # inputs: Video as tensor (b, c, f, h, w)
+    # outputs: Encodings as tensor (b, 4, h/8, w/8)
+    def encode_video_latents(self,vid):
+        if vid.dim != 5:
+            vid_array = vid[None,:]
+        video_by_batch = vid_array.float().permute(0, 2, 1, 3, 4)
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:72]: Note: Change frame and channel order (from [b, c, f, h, w] to [b, f, c, h, w]).

File: mindglass_code/autoencoder.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        pil_images = [Image.fromarray(image) for image in imgs]
+        return pil_images
+
+    # Define function that encodes a video
+    # Encodes videos to the latent space.
+    # Can pass in videos of any size - all dims recuded by a factor of 8
+    # inputs: Video as tensor (b, c, f, h, w)
+    # outputs: Encodings as tensor (b, 4, h/8, w/8)
+    def encode_video_latents(self,vid):
+        if vid.dim != 5:
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:70]: Assuming the video comes as a tensor, it should be: `vid.dim()`

File: mindglass_code/autoencoder.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            vid_array = vid[None,:]
+        video_by_batch = vid_array.float().permute(0, 2, 1, 3, 4)
+        if torch.cuda.is_available():
+            video_by_batch = video_by_batch.cuda()
+        latent_samples = []
+        for i in range (video_by_batch.shape[0]):
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:76]: Note: This loops through the batches.

File: mindglass_code/autoencoder.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            latent_dists_1_video = self.model.encode(video_by_batch[i])
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:77]: `video_by_batch[i]` still has shape [f, c, h, w], I believe it requires a cascaded loop through the frames.

File: mindglass_code/autoencoder.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            latent_sample = latent_dists_1_video.latent_dist.sample()
+            latent_samples.append(latent_sample)
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:79]: Not sure if this creates the correct stacked dimension, I would rather use `torch.stack`: https://pytorch.org/docs/stable/generated/torch.stack.html

Comment (oxjohanndiep) [reply mindglass_code/autoencoder.py:79]: And stack it at dimension 1 such that the end dimension would be [b, f, 4, h/8, w/8)].

File: mindglass_code/autoencoder.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        imgs = imgs.detach().cpu().permute(0, 2, 3, 1).numpy()
+        imgs = (imgs * 255).round().astype('uint8')
+        pil_images = [Image.fromarray(image) for image in imgs]
+        return pil_images
+
+    # Define function that encodes a video
+    # Encodes videos to the latent space.
+    # Can pass in videos of any size - all dims recuded by a factor of 8
+    # inputs: Video as tensor (b, c, f, h, w)
+    # outputs: Encodings as tensor (b, 4, h/8, w/8)
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:68]: Should be [b, f, 4, h/8, w/8], I believe you forgot the stacked frame embeddings per batch.

File: mindglass_code/autoencoder.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        return latent_samples
+
+    # Decodes video latents
+    # inputs: latent encoding of video as tensor (b, f, dc, dh, dw)
+    # outputs: video as tensor (b, c, f, h, w)
+    def decode_video_latents(self, latents):
+        latents = 1 / 0.18215 * latents
+        videos = []
+        for i in range(latents.shape[0]):
+            with torch.no_grad():
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:93]: Are you sure e need `torch.no_grad()` only in the decoder?

File: mindglass_code/autoencoder.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+                video_by_batch = self.model.decode(latents[0])
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:94]: Same as above, I believe it requires a cascaded for-loop through the frame-dimension.

File: mindglass_code/autoencoder.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+            videos.append(video_by_batch[0])
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:95]: Same as above, I would use `torch.stack`.

File: mindglass_code/autoencoder.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+        video_tensor = torch.stack(videos).permute(0, 2, 1, 3, 4)
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:97]: Yes using this but without the appending to an empty array.

**Review by oxjohanndiep — CHANGES_REQUESTED** [review]
Went through the first review, request some changes. It would be nice if you can test it on mock data first, both the image and video encoder/decoder functions. This way you can be 100% about the individual dimensions.

File: mindglass_code/autoencoder.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+        self.model = AutoencoderKL.from_pretrained('CompVis/stable-diffusion-v1-4', subfolder='vae', use_auth_token=True)
+        if torch.cuda.is_available():
+                self.model = self.model.cuda()
+
+    # Define function that encodes the image 
+    # Encodes images to the latent space.
+    # Can pass in images of any size - all dims recuded by a factor of 8
+    # inputs: Images (b, h, w)
+    # outputs: Encodings(b, 4, h/8, w/8)
+    def encode_img_latents(self,imgs):
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:30]: Why do we need this method? Are we gonna use it?

File: mindglass_code/autoencoder.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+        latent_samples = latent_dists.latent_dist.sample()
+        latent_samples *= 0.18215
+
+        return latent_samples
+
+    # Decodes images to the latent space.
+    # Can pass in images of any size - all dims recuded by a factor of 8
+    # inputs: (b, dh, dw)
+    # outputs: (b, 4, dh*8, dw*8)
+    def decode_img_latents(self, latents):
```
Comment (oxjohanndiep) [inline mindglass_code/autoencoder.py:51]: Why do we need this method, are we gonna use it?

**Review by oxjohanndiep — COMMENTED** [review]
Minor comments regarding your two methods for encoding and decoding images. I believe you do not even use it at the end, as for video encoding/decoding you are calling directly the Huggingface class.

File: mindglass_code/autoencoder.py (hunk 6 above)
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:77]: The .encode function handles batches of inputs. So I believe that the function will handle the frames (as if a batch of images) and then the for loop handles a batch of video inputs.

File: mindglass_code/autoencoder.py (hunk 2 above)
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:59]: My comment was only for line 58

File: mindglass_code/autoencoder.py (hunk 13 above)
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:30]: No! Guess it is just the legacy code

File: mindglass_code/autoencoder.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+        # Unsure why this resizing is done
+        img_arr = img_arr / 255.0
+        img_arr = torch.from_numpy(img_arr).float().permute(0, 3, 1, 2)
+        img_arr = 2 * (img_arr - 0.5)
+        if torch.cuda.is_available():
+            img_arr = img_arr.cuda()
+        latent_dists = self.model.encode(img_arr)
+        latent_samples = latent_dists.latent_dist.sample()
+        latent_samples *= 0.18215
```
[the remark this answers is not in the document]
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:43]: Unsure - will look into this further

File: mindglass_code/autoencoder.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+        return latent_samples
+
+    # Decodes images to the latent space.
+    # Can pass in images of any size - all dims recuded by a factor of 8
```
[the remark this answers is not in the document]
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:48]: Have corrected!!

File: mindglass_code/autoencoder.py (hunk 14 above)
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:51]: Again, no but is the base code

File: mindglass_code/autoencoder.py (hunk 9 above)
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:93]: I guess we need it everywhere as we wont be updating weights of this model or of any of the inputs.

File: mindglass_code/autoencoder.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+class VAE_pretrained():
+    def __init__(self):
+        self.model = AutoencoderKL.from_pretrained('CompVis/stable-diffusion-v1-4', subfolder='vae', use_auth_token=True)
+        if torch.cuda.is_available():
+                self.model = self.model.cuda()
+
+    # Define function that encodes the image 
+    # Encodes images to the latent space.
+    # Can pass in images of any size - all dims recuded by a factor of 8
+    # inputs: Images (b, h, w)
```
[the remark this answers is not in the document]
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:28]: Update comment accordingly

File: mindglass_code/autoencoder.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+    # outputs: Encodings(b, 4, h/8, w/8)
+    def encode_img_latents(self,imgs):
+        if not isinstance(imgs, list):
+            imgs = [imgs]
+        img_arr = np.stack([np.array(img) for img in imgs], axis=0)
+
+        # Unsure why this resizing is done
+        img_arr = img_arr / 255.0
+        img_arr = torch.from_numpy(img_arr).float().permute(0, 3, 1, 2)
+        img_arr = 2 * (img_arr - 0.5)
```
[the remark this answers is not in the document]
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:38]: Thanks for this! Will have a look at how this translates to video inputs. I think we will not have to account for this as it is handled in the gif_to_tensor function

File: mindglass_code/autoencoder.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+        latent_dists = self.model.encode(img_arr)
+        latent_samples = latent_dists.latent_dist.sample()
+        latent_samples *= 0.18215
+
+        return latent_samples
+
+    # Decodes images to the latent space.
+    # Can pass in images of any size - all dims recuded by a factor of 8
+    # inputs: (b, dh, dw)
+    # outputs: (b, 4, dh*8, dw*8)
```
[the remark this answers is not in the document]
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:50]: Updates

File: mindglass_code/autoencoder.py (hunk 1 above)
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:58]: Updated

File: mindglass_code/autoencoder.py (hunk 8 above)
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:68]: Updated

File: mindglass_code/autoencoder.py (hunk 4 above)
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:70]: Updated

File: mindglass_code/autoencoder.py (hunk 7 above)
Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:79]: I do not think this will work without an if condition. Given the first iteration of the for loop there is not a base tensor with which we want to stack.

Comment (Ljackson-su) [reply mindglass_code/autoencoder.py:79]: I am unsure how this will work since on the first iteration of the for loop there is not a base tensor to use

File: mindglass_code/autoencoder.py (hunk 6 above)
Comment (oxjohanndiep) [reply mindglass_code/autoencoder.py:77]: I see, smart choice!

File: mindglass_code/autoencoder.py (hunk 7 above)
Comment (oxjohanndiep) [reply mindglass_code/autoencoder.py:79]: Trying with some mock-data, this approach works great!

**Comment by oxjohanndiep** [comment]
Accidently pre-merged this `feature_1` into `develop`, which is why this pull-request does not contain the new commits to `feature_1`. The merge into `develop` has now been closed with a hotfix.
